Add one-sided local subcell IDP limiting for non-linear variables

bennibolm opened the pull request:
The positivity limiting for non-linear variables was merged ([PR](https://github.com/trixi-framework/Trixi.jl/pull/1738)) and therefore the Newton-bisection method is now included to `main`.
So now, the Newton method can easily be used for a ***local one-sided limiting of nonlinear variables***.
General variables, e.g. `entropy_guermond_etal` or `entropy_math`, can be passed combined with a specification of the wanted bound, i.e. pass `min` for a lower bound and `max` for upper bounds.
Example: Pass the following parameter within the limiter
```
local_onesided_variables_nonlinear = [(Trixi.entropy_guermond_etal, min)]
```

Additional, in this PR I renamed `local_minmax` to `local_twosided` to be consistent with `local_onesided`.

- [x] Generalization from limiting spec and math entropy to arbitrary variables in https://github.com/bennibolm/Trixi.jl/pull/125

bennibolm commented on src/equations/compressible_euler_2d.jl:
```diff
@@ -1588,6 +1588,33 @@ end
+    w3 = -rho_v2 * inv_rho_gammap1
+    w4 = (1 / rho)^equations.gamma
+
+    # The derivative vector for other specific entropy
+    # sp = 1.0/(gammam1 * (rho_e - 0.5 * rho * v_square)
+    # w1 = gammam1 * 0.5 * v_square * sp - gamma / rho
+    # w2 = -gammam1 * v1 * sp
+    # w3 = -gammam1 * v2 * sp
+    # w4 = gammam1 * sp
```
I'm not sure if it makes sense to keep this comment. I copied it from Fluxo back then.
Do you use it sometimes? @amrueda

amrueda replied:
Good question!
We have mainly used this "modified" specific entropy (`s = e * rho^(1 - gamma) = p *rho^(-gamma) / (gamma - 1)`) because it is very well suited for the Newton's method and guarantees that the "conventional" specific entropy (`s = ln(p / rho^gamma)`) of the Euler equations is within bounds. I would definitely add that to the description of the subroutine, as suggested above.

If you are not in the mood for creating new test cases for the "conventional" specific entropy, I think that it is fine if you just remove the comments. 
```suggestion
```

bennibolm replied:
Done in [ad7e1da](https://github.com/trixi-framework/Trixi.jl/pull/1792/commits/ad7e1dabf3c532d18b8af45b0b490b47b4a4e3ef)

amrueda commented on src/equations/compressible_euler_2d.jl:
```diff
@@ -1854,6 +1854,33 @@ end
     return SVector(w1, w2, w3, w4)
 end
 
+# Transformation from conservative variables u to entropy vector dSdu,
+# S = -rho*s/(gamma-1), s=ln(p)-gamma*ln(rho)
```
That `S` is the mathematical entropy and not really the specific entropy that this function uses. I suggest the following change:
```suggestion
# Transformation from conservative variables u to entropy vector ds_0/du,
# using the specific entropy of Guermond et al. (2019): s_0 = p * rho^(-gamma) / (gamma-1).
# Note: This is *not* the "conventional" specific entropy s = ln(p / rho^(gamma)).
```
However, we can discuss if we want to add more information.

bennibolm replied:
I think I just copied the explanation of the wrong entropy from Fluxo when I added this routine. For me, it's good, as you suggested.

bennibolm replied:
Done in [ad7e1da](https://github.com/trixi-framework/Trixi.jl/pull/1792/commits/ad7e1dabf3c532d18b8af45b0b490b47b4a4e3ef)

amrueda commented on src/equations/compressible_euler_2d.jl:
```diff
@@ -1959,6 +1986,31 @@ end
+    # Modified specific entropy from Guermond et al. (2019)
+    s = (rho_e - 0.5 * (rho_v1^2 + rho_v2^2) / rho) * (1 / rho)^equations.gamma
+
+    # Other specific entropy
+    # rho_sp = rho/((equations.gamma - 1.0) * (rho_e - 0.5 * rho * v_square))
+    # s = log(p) - (equaions.gamma + 1) * log(rho)
```
We should then remove this if not used.
```suggestion
```

bennibolm replied:
Done in [ad7e1da](https://github.com/trixi-framework/Trixi.jl/pull/1792/commits/ad7e1dabf3c532d18b8af45b0b490b47b4a4e3ef)

amrueda reviewed (changes requested):
Thanks @bennibolm!
This looks very good to me. Here a couple of comments:

bennibolm commented on src/callbacks_stage/subcell_bounds_check_2d.jl:
```diff
@@ -43,9 +43,30 @@
+                for j in eachnode(solver), i in eachnode(solver)
+                    v = variable(get_node_vars(u, equations, solver, i, j, element),
+                                 equations)
+                    if min_or_max === max
+                        deviation = max(deviation,
+                                        v - variable_bounds[key][i, j, element])
+                    else # min_or_max === min
+                        deviation = max(deviation,
+                                        variable_bounds[key][i, j, element] - v)
+                    end
```
Good idea, thanks. I now moved the min/max calculation in front of the `for` loop but took your version otherwise.

bennibolm replied:
Done in [ad7e1da](https://github.com/trixi-framework/Trixi.jl/pull/1792/commits/ad7e1dabf3c532d18b8af45b0b490b47b4a4e3ef).

bennibolm commented on src/callbacks_stage/subcell_bounds_check_2d.jl:
```diff
@@ -43,9 +43,30 @@
         end
     end
+    if local_onesided
+        foreach(limiter.local_onesided_variables_nonlinear) do (variable, min_or_max)
```
This is coming from [this issue with allocations](https://github.com/bennibolm/Trixi.jl/pull/125#discussion_r1506188560). Though, after the last fix in that PR (create a Tuple out of the vector for `local_onesided_variables_nonlinear`) the foreach loops are possibly redundant. I will check this.

bennibolm replied:
I replaced `foreach` by `for` in [27a93f4](https://github.com/trixi-framework/Trixi.jl/pull/1792/commits/27a93f402d0cdeda25dd77d348e08b3add87c9a0)

sloede commented on src/equations/compressible_euler_2d.jl:
```diff
@@ -1886,6 +1886,27 @@ end
+# Transformation from conservative variables u to entropy vector ds_0/du,
+# using the specific entropy of Guermond et al. (2019): s_0 = p * rho^(-gamma) / (gamma-1).
+# Note: This is *not* the "conventional" specific entropy s = ln(p / rho^(gamma)).
+@inline function cons2entropy_spec(u, equations::CompressibleEulerEquations2D)
```
Just noticed this... sorry: We now have `cons2entropy_spec` but it's _not_ the specific entropy. Isn't this bound to someone making an error and using the wrong one? Should we maybe just name this `cons2entropy_guermondetal`?

bennibolm replied:
That's actually a good point. I think that come from Fluxo where this routine originally used the real specific entropy but was changes later on. I'm fine with changing it.
Would you still keep the additional note here and below?

sloede replied:
> Would you still keep the additional note here and below?

Absolutely!

bennibolm replied:
Thanks. Done in [27ded75](https://github.com/trixi-framework/Trixi.jl/pull/1792/commits/27ded75f469a66d6632815626d373549710bf1fc)

sloede commented on src/equations/compressible_euler_2d.jl:
```diff
@@ -1991,6 +2012,28 @@ end
+end
+
+# Calculate specific entropy for conservative variable u
+@inline function entropy_spec(u, equations::CompressibleEulerEquations2D)
```
```suggestion
@inline function entropy_guermondetal(u, equations::CompressibleEulerEquations2D)
```
As above

bennibolm replied:
[27ded75](https://github.com/trixi-framework/Trixi.jl/pull/1792/commits/27ded75f469a66d6632815626d373549710bf1fc)

sloede commented on src/equations/compressible_euler_2d.jl:
```diff
@@ -1991,6 +2012,28 @@ end
+    return cons2entropy(u, equations)
+end
+
+# Calculate specific entropy for conservative variable u
```
```suggestion
# Calculate the specific entropy of Guermond et al. (2019): s_0 = p * rho^(-gamma) / (gamma-1).
# Note: This is *not* the "conventional" specific entropy s = ln(p / rho^(gamma)).
```
I think it wouldn't hurt to repeat this comment here

bennibolm replied:
Done in [27ded75](https://github.com/trixi-framework/Trixi.jl/pull/1792/commits/27ded75f469a66d6632815626d373549710bf1fc)

sloede commented on src/solvers/dgsem_tree/subcell_limiters.jl:
```diff
@@ -71,32 +83,56 @@ end
+    local_onesided_variables_nonlinear_ = Tuple{Function, Function}[]
+    for (variable, min_or_max) in local_onesided_variables_nonlinear
+        if min_or_max === Base.max
+            push!(local_onesided_variables_nonlinear_, tuple(variable, max))
+        elseif min_or_max === Base.min
+            push!(local_onesided_variables_nonlinear_, tuple(variable, min))
+        elseif min_or_max === Trixi.max || min_or_max === Trixi.min
+            push!(local_onesided_variables_nonlinear_, tuple(variable, min_or_max))
```
```suggestion
            push!(local_onesided_variables_nonlinear_, (variable, max))
        elseif min_or_max === Base.min
            push!(local_onesided_variables_nonlinear_, (variable, min))
        elseif min_or_max === Trixi.max || min_or_max === Trixi.min
            push!(local_onesided_variables_nonlinear_, (variable, min_or_max))
```
Don't think `tuple` is necessary here, is it?

bennibolm replied:
Done in [62946c2](https://github.com/trixi-framework/Trixi.jl/pull/1792/commits/62946c2220b1005ac1ff9e8976425590d22e27e9)

sloede reviewed (changes requested):
Almost there 😊

sloede reviewed (approved):
LGTM!

sloede commented:
All failures seem to be Codecov related, but we still cannot see code coverage 😢

bennibolm commented:
> All failures seem to be Codecov related, but we still cannot see code coverage 😢

If it now fails again, I will rerun the failed tests overnight. That works most of the time.

bennibolm commented:
@sloede All tests have been run through now. The codecov tests still slightly fail, mostly due to not tested line within the saving deviations part of the BoundsCheckCallback.
Is it okay like this? Otherwise, I would have to enable more limiters in the one or two elixirs where the deviations are actually saved in the tests.

sloede commented:
It seems to me that I/O for one-sided or two-sided limiters is not covered in the tests. If I were you, I'd make sure to have this part executed at least in one test to ensure it will continue to work even if you decide to modify some variables somewhere else.

But I'll leave the decision up to you on how to proceed.

bennibolm commented:
> It seems to me that I/O for one-sided or two-sided limiters is not covered in the tests. If I were you, I'd make sure to have this part executed at least in one test to ensure it will continue to work even if you decide to modify some variables somewhere else.
> 
> But I'll leave the decision up to you on how to proceed.

I already included a two sided limiter I/O test in https://github.com/trixi-framework/Trixi.jl/pull/1824. So, I think it would merge this PR as it is and handle the one sided I/O in the other PR (or a completely new one).

sloede commented:
> I already included a two sided limiter I/O test in #1824.

Are you sure? On Codecov, I am seeing this
<img width="766" alt="image" src="https://github.com/trixi-framework/Trixi.jl/assets/3637659/05f94edc-7c6b-44db-8714-fd48d97ee31a">

for `src/callbacks_stage/subcell_bounds_check_2d.jl`, and this
<img width="768" alt="image" src="https://github.com/trixi-framework/Trixi.jl/assets/3637659/63546205-8dba-415f-bba6-cfe98c6b171b">

for `src/callbacks_stage/subcell_bounds_check.jl`.

But if that's ok, we can still merge right away 😊

bennibolm commented:
> > I already included a two sided limiter I/O test in #1824.
> 
> Are you sure? On Codecov, I am seeing this <img alt="image" width="766" src="https://private-user-images.githubusercontent.com/3637659/328851896-05f94edc-7c6b-44db-8714-fd48d97ee31a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNjQxOTEsIm5iZiI6MTcxNTE2Mzg5MSwicGF0aCI6Ii8zNjM3NjU5LzMyODg1MTg5Ni0wNWY5NGVkYy03YzZiLTQ0ZGItODcxNC1mZDQ4ZDk3ZWUzMWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDhUMTAyNDUxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjFmZDEzMWJlYTg2MmNjYmMxMjM1YWRhZTE5NTIzZTc1OTRlYjRkZTNkMDQ5MmYzZGYyOTE2NmM4MjViMjY1YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Jvf7sCFLToapsUck8Ftejw7w4uXUr6SjBMbPux2POfM">
> 
> for `src/callbacks_stage/subcell_bounds_check_2d.jl`, and this <img alt="image" width="768" src="https://private-user-images.githubusercontent.com/3637659/328852105-63546205-8dba-415f-bba6-cfe98c6b171b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNjQxOTEsIm5iZiI6MTcxNTE2Mzg5MSwicGF0aCI6Ii8zNjM3NjU5LzMyODg1MjEwNS02MzU0NjIwNS04ZGJhLTQxNWYtYmJhNi1jZmU5OGM2YjE3MWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDhUMTAyNDUxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDkwZjhhZWIwNzdhYjU5Y2ZlOTZhNDIyZjE3YWE4OWQyNTA4OTNjYWU0YzI5YjQzOTRiMDYxODcxODgyZTA5OSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.sXl6Q5xK2MyEcYh9O-fmieK5YswEUOpYvM6iYrfvx8c">
> 
> for `src/callbacks_stage/subcell_bounds_check.jl`.
> 
> But if that's ok, we can still merge right away 😊

Oh, you're right. I just added the test for `elixir_euler_sedov_blast_wave_sc_subcell` without any additional limiter. But I added the onesided limiter to this specific elixir in this PR here, so I'll have to deal with it anyway :sweat_smile: 

So, I would go for it and merge it :rocket:
